### ISSUE TITLE
Does itemComponent work for you?

### ISSUE THREAD

**Issue by karellm** [miguelcobain/ember-cli-selectize#135]
Hi,

I'm building an app with Ember 1.13.x and tried to use `itemComponent` without success. I tried to debug it and came down to that part of the app: https://github.com/miguelcobain/ember-cli-selectize/blob/master/addon/components/ember-selectize.js#L787

It basically is an empty component. If I add `classNames: ['bla']` for instance, it is added to the component but the template is simply never rendered.

I think it is because Glimmer changed that behavior. You can see this here for instance: https://60devs.com/hacky-way-to-render-ember-template-to-string-in-ember-2-0.html

Before diving deeper I wanted to make sure I didn't miss anything. Thanks!

**Comment by miguelcobain** [comment]
The test is passing. :/
I'll confirm in a minute. What is your ember version?

**Comment by miguelcobain** [comment]
Also, what selective version is on your bower package?

**Comment by karellm** [comment]
From what I can say, you do not test the `itemComponent` feature (I skimmed the tests w/o success).

I just upgraded to ember-cli-selectize 0.5.1 and run ember 1.13.12

**Comment by miguelcobain** [comment]
The problem is probably the selectize bower version. Please run `ember g
ember-cli-selectize` to install the new selectize version.

A ter, 19/01/2016, 19:50, Karel Ledru-Mathe <notifications@github.com>
escreveu:

**Comment by karellm** [comment]
It requires master and I'm not favorable to run out of master. It is too risky for a production app. Is there a reason for that?

**Comment by karellm** [comment]
Also at this point, it hasn't got to selectize per say. The component being empty really is a `ember-cli-selectize` issue I think.

**Comment by karellm** [comment]
I think the upgrade really caused me more trouble than it solved. Now the `x` cross does not work. Though the backspace keypress works.

**Comment by miguelcobain** [comment]
@karellm pulling from master is a temporary workaround.

tl;dr selectize merged in a PR that lets render functions to accept DOM elements (and not only strings).
However, that PR wasn't released yet on selectize repo. So, I had to build it myself and publish it temporarily.

So, this master version is necessary to be able to render components this way.

**Comment by karellm** [comment]
@miguelcobain I did the update. The item cross to remove items stopped working. 

Can you confirm me that `itemComponent` works for you with a component and separate hbs template? Also can you show where you test that? Thanks!

**Comment by karellm** [comment]
@miguelcobain The commit you made fixes the `_componentToDOM` method. It is no more empty :+1: 

That said, you test for `optionComponent`, not `itemComponent`. I still get the same error from selectize.js: `Assertion failed: TypeError: html_container.search is not a function(…)`. I'm wondering if it is not because you return the DOM structure rather than a string. So there is obviously no `search` method available.

**Comment by miguelcobain** [comment]
> I'm wondering if it is not because you return the DOM structure rather than a string. So there is obviously no search method available.

@karellm That's exactly why I insisted that you must use my fork of selectize. That fork includes a PR that allows render functions to return DOM instead of strings.
With this, you still get bindings normally, something that isn't possible with strings.

Please double check that you're running the correct selectize version (show me your bower.json version) and reopen if the issue persists.

**Comment by karellm** [comment]
Sorry, I was working on another branch and forgot to re-run `bower install` when testing this morning. I confirm that it fixes the issue I mentioned above. Thanks!

Last thing, does the cross to remove an item work for you though? I can remove item using the keyboard delete actions but not by clicking the items' cross.

**Comment by miguelcobain** [comment]
@karellm you mean that the crosses don't appear at all in *multiple* mode?

**Comment by miguelcobain** [comment]
@karellm 
The remove button is a core selectize **plugin**. Unfortunately, that is how selectize works.

You can see here that the plugin overrides the item render function: https://github.com/selectize/selectize.js/blob/master/src%2Fplugins%2Fremove_button%2Fplugin.js#L46-L51

So,
- ideal solution - solve this problem in the selectize plugin itself and PR selectize
- workaround you probably can mimic that plugins behavior in your own item render function

**Comment by karellm** [comment]
It does appear and is clickable but the action does nothing and returns a:

```
Uncaught TypeError: Cannot read property 'nodeType' of undefined
jQuery.fn.extend.removeClass @ jquery.js:7230
$.extend.setActiveItem @ selectize.js:1855
$.extend.onBlur.deactivate @ selectize.js:1683
$.extend.onBlur @ selectize.js:1699
$.extend.setup.$control_input.on.blur @ selectize.js:1251
jQuery.event.dispatch @ jquery.js:4435
jQuery.event.add.elemData.handle @ jquery.js:4121
```

**Comment by karellm** [comment]
Looking at `selectize` on github, the project seems dead or dying. Writing a PR to a project with 76 open PRs feels pointless. Especially since it won't work with your custom version, meaning that at this point I can't use your addon anymore... If you can take a stab at it within the addon, it would be very appreciated.

Otherwise I guess I will do like everyone else: http://stackoverflow.com/a/13575921/999540.

Thanks for you help on the matter. It is pleasant to see people caring for their OS projets :)

**Comment by miguelcobain** [comment]
@karellm If you're starting out, you should use http://www.ember-power-select.com/
I'm also involved in the project.

**Comment by atremblay-seniorlink** [comment]
This same error is happening for me with Ember 2.4.2 and bower 1.7.7 with ember-cli-selectize set to 0.5.5 and selectize set to `miguelcobain/selectize.js#master` in my bower.json. It only occurs when trying to set itemComponent multiple (optionComponent works fine). I traced it to an anonymous function inside the `Selectize.define('remove_button'` function where `search()` was being called on an object rather than a string. 

It would probably be fairly straightforward to implement a check to ignore that specific Selectize define function when a custom template is provided and just add the required event listener manually. Or put in a shim for the passed object to implement `search()`. 

...but if @miguelcobain is moving to another project maybe it would be safer not to use this repo at all?

**Comment by karellm** [comment]
@atremblay-seniorlink We were using a combination of `ember-cli-selectize`, `ember-x-select` and `Ember.Select` and the transition took some time but was definitely worth it. I can ensure you that `ember-power-select` will likely cover all the use cases you may think of but also sometimes require a little more coding.

In our project we had recurring use cases and wrote wrapper components for these that use `ember-power-select` under the hood. It gave use a syntax that is very close to `Ember.Select` for simple selects, yet using always a single library.

**Comment by miguelcobain** [comment]
@atremblay-seniorlink I'm not "moving" to another project.

tl;dr is
- using a pure ember select solution like ember-power-select will get you better performance and you get other benefits like smaller payload sizes and better fastboot support
- nevertheless this project is still maintained. there are many people using it. If you're starting out, I recommend ember-power-select
